### PR TITLE
PSA crypto should not depend on the cipher module

### LINKED ISSUE #8048: [G2] PSA crypto should not depend on the cipher module
As of Mbed TLS 3.4, `MBEDTLS_PSA_CRYPTO_C` is registered as depending on `MBEDTLS_CIPHER_C` in `check_config.h`. This is obsolete, dating back from before we had driver support for ciphers. Now, if there are no cipher/AEAD algorithms that use the built-in implementation, PSA doesn't need `MBEDTLS_CIPHER_C`.

At the time of writing, the build-in implementation of all non-authenticated ciphers depends heavily on the cipher module. On the other hand, the build-in implementation of AEAD only directly depends on cipher for metadata conversion. Then the built-in GCM and CCM modules themselves depend on cipher, but chachapoly doesn't. So it should be easy to make a cipher-less chachapoly-only build possible.

The goal of this task is:

* In `psa_aead_setup`, only call `mbedtls_cipher_info_from_psa` if `MBEDTLS_CIPHER_C` is defined.
* Change `check_config.h` to only require `MBEDTLS_CIPHER_C` based on `MBEDTLS_PSA_CRYPTO_C` if the built-in implementation of one of (a non-authenticated cipher, CCM, GCM) is enabled (`PSA_HAVE_SOFT_CIPHER || MBEDTLS_PSA_BUILTIN_ALG_CCM || MBEDTLS_PSA_BUILTIN_ALG_GCM`, where `PSA_HAVE_SOFT_CIPHER` needs to be defined in `config_psa.h` or its successor).
* Make sure that when `MBEDTLS_PSA_CRYPTO_CONFIG` is enabled, `config_psa.h` or its successor enables `MBEDTLS_CIPHER_C` when required.
* In `component_test_full_no_cipher`, change `test_full_no_cipher` to keep PSA enabled.

Follow-up:

* https://github.com/Mbed-TLS/mbedtls/issues/6317: add or change a build with all ciphers accelerated to have `MBEDTLS_CIPHER_C` disabled.

### PR DESCRIPTION
**Pull request by valeriosetti**
Resolves #8048 

~Depends on #8315 (PR #8368)~

## PR checklist

- [x] **changelog** deferred to #8357
- [x] **backport** not required
- [x] **tests** not required, no new code introduced

### CODE REVIEW

**Comment by valeriosetti** [comment]
The last commit (https://github.com/Mbed-TLS/mbedtls/pull/8119/commits/41a39718a22d945c60d988dc32eae1d085cda7e1) implements what has been proposed in the [original task](https://github.com/Mbed-TLS/mbedtls/issues/8048#issuecomment-1692826911)

File: tests/scripts/all.sh
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:1509]: It should be possible to keep `MBEDTLS_PSA_CRYPTO_SE_C` and `MBEDTLS_PSA_CRYPTO_STORAGE_C` now.

**Review by gabor-mezei-arm — APPROVED** [review]
Looks good to me

**Comment by valeriosetti** [comment]
I just rebased this in order to solve conflicts with `development`

**Review by mpg — COMMENTED** [review]
Just a first pass. Left a few thoughts and questions. The main point is about testing: I think we want to leave the existing component alone and add a new one instead.

File: include/psa/crypto_builtin_primitives.h
Comment (valeriosetti) [reply include/psa/crypto_builtin_primitives.h:102]: Yes I thought the same once I read the first comment so I'm setting the dependency now in the description

File: include/mbedtls/config_adjust_legacy_from_psa.h
Comment (valeriosetti) [reply include/mbedtls/config_adjust_legacy_from_psa.h:701]: Right. I will drop this commit

File: include/psa/crypto_builtin_primitives.h
Comment (valeriosetti) [reply include/psa/crypto_builtin_primitives.h:102]: Dropping this commit as well

**Comment by valeriosetti** [comment]
Since the original PR was designed before the plan for ciphers/AEAD was fully developed, I made some changes to the commits' structure to (hopefully) make this PR clearer:

- now it depends on #8315 so the initial commits come from there
- it also addresses the first review's comments from @mpg 

So far it only misses this part from [this comment](https://github.com/Mbed-TLS/mbedtls/pull/8119#discussion_r1360469909):
> or perhaps even two of them: one with PSA_CRYPTO_CONFIG and one without?

I'm going to work on this as follow up commit

**Comment by mpg** [comment]
> So far it only misses this part from [this comment](https://github.com/Mbed-TLS/mbedtls/pull/8119#discussion_r1360469909):
> 
> > or perhaps even two of them: one with PSA_CRYPTO_CONFIG and one without?
> 
> I'm going to work on this as follow up commit

Note: in the meantime I convinced myself that it's enough to test `PSA_CRYPTO_CONFIG` enabled.

**Comment by valeriosetti** [comment]
> > So far it only misses this part from [this comment](https://github.com/Mbed-TLS/mbedtls/pull/8119#discussion_r1360469909):
> > > or perhaps even two of them: one with PSA_CRYPTO_CONFIG and one without?
> > 
> > 
> > I'm going to work on this as follow up commit
> 
> Note: in the meantime I convinced myself that it's enough to test `PSA_CRYPTO_CONFIG` enabled.

In the end I added 2 components, one with only `CRYPTO_C` and the 2nd with also `CRYPTO_CONFIG`:
- component_test_full_no_cipher_with_crypto()
- component_test_full_no_cipher_with_crypto_config()

The 2nd is what you asked for in the comment above. But I think that also the 1st one makes sense since it includes a fix for `adjust_psa_from_legacy` (similar to the one already done in the parent PR #8368 for `adjust_legacy_from_psa`) which otherwise would have been skipped.
@mpg wdyt? If you think this is not necessary and it just adds extra work to the CI, I can remove the 1st component.

**Comment by mpg** [comment]
I think if the first component uncovered issues that the 2nd didn't, then it's a clear sign I was wrong and we need both of them :)

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:1545]: I don't find this comment very clear now. Suggestion: "The built-in implementation of these modes currently depend on cipher.c."

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:1565]: Suggestions: "The following modules directly depends on CIPHER_C".

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:1572]: Note: with `USE_PSA_CRYPTO`, this should not depend on `CIPHER_C`, but that's our of scope and will be another issue.

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:1592]: Really? Why? (Same for LMS below.)

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:1585]: Pre-existing: I think that's a big vague. I'd rather have several paragraphs like "These depend on SSL_TLS_C", "These depend on XXX", etc.

I think if we had that in `full_no_cipher` then `USE_PSA` and LMS would have been in the paragraph "These depend on `PSA_CRYPTO_C`, making it apparent that we don't need to disable them here - unless I'm missing something.

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:1605]: Same note regarding `-` (and below as well).

**Review by mpg — CHANGES_REQUESTED** [review]
This is going in the right direction, but there are a couple issues, and the CI is not fully happy yet - looks like more "not supported" vs "invalid" issues.

File: tests/scripts/all.sh
Comment (valeriosetti) [reply tests/scripts/all.sh:1592]: That was inherited by the original `full_no_cipher` component. In that one `CRYPTO_C` is disabled so `USE_PSA_CRYPTO` was as well due to its dependency on the former. You're right, we don't need that in this case, so I will re-enable it.

File: tests/scripts/all.sh
Comment (mpg) [reply tests/scripts/all.sh]: Actually could you also tidy the "Indirect dependencies" section in the existing `full_no_cipher` component?
- I believe most TLS stuff there no longer needs to be present as it now get auto-disabled when `SSL_TLS_C` is disabled (which is done in the "direct dependencies" section);
- For the others, I'd like to have a specific justification for each (group): "depends on PSA_CRYPTO_C", "depends on XXX", etc.

Comment (valeriosetti) [reply tests/scripts/all.sh]: Oh, sorry, I thought that I had just to clean the new component I added, but of course it's better to improve comments also on the original one. I will do that ASAP

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:1574]: Actually I don't think that's the right place for those: IMO they should be in the "else" part of the `if [ "$USE_CRYPTO_CONFIG" -eq 1 ]` above: when using PSA-based config, we need to disable PSA symbols, and when using legacy-based config, we need to disable legacy symbols (otherwise, the corresponding PSA features would get enabled, making PSA depend on cipher).

Or am I missing some way the legacy options still influence PSA options even when `PSA_CRYPTO_CONFIG` is enabled?

**Review by mpg — CHANGES_REQUESTED** [review]
Thanks for addressing my feedback! I'm happy with the rebase, but I found one more thing while looking again at the code.

File: tests/scripts/all.sh
Comment (valeriosetti) [reply tests/scripts/all.sh:1574]: I'm not sure: `mbedtls_config.h` is evaluated independently of the status of `PSA_CRYPTO_CONFIG`. The `adjust` header files (either legacy->PSA or PSA->legacy) are additive, not subtractive. So if for example we have `MBEDTLS_CCM_C` enabled in `mbedtls_config.h` then it will still be enabled even if `PSA_CRYPTO_CONFIG` is evaluated and `PSA_WANT_ALG_CCM` is disabled. 
Am I wrong?

**Comment by valeriosetti** [comment]
As anticipated [here](https://github.com/Mbed-TLS/mbedtls/pull/8119#issuecomment-1775158699) I rebased the PR in order to solve conflict with `development` after #8368 was merged

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:1525]: Ah, just noticed: I'd prefer `with_psa_crypto`, just "crypto" seems less clear, considering we still have non-PSA crypto.

File: tests/scripts/all.sh
Comment (mpg) [reply tests/scripts/all.sh:1564]: For CCM and GCM, yes, sure, but they're already disabled in the section "depens on `CIPHER_C`" as they should.

Here I was talking about "cipher modes/keys that make PSA depend on CIPHER_C" as the comment says. My point is:
- we don't need to disable these in the `full_no_cipher` component that doesn't have PSA;
- so we only disable them because of the influence they could have on PSA;
- unless I'm missing something, they should only influence PSA when `PSA_CRYPTO_CONFIG` is disabled;
- so, we should only need to disable them when `PSA_CRYTO_CONFIG` is disabled.

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:1535]: Minor: we should add a note that it's currently everything except Chacha20 and Chacha20-Poly1305.

**Review by mpg — COMMENTED** [review]
Just checking the rebase, with which I'm happy, and requesting more comment improvements while at it :)

File: tests/scripts/all.sh
Comment (valeriosetti) [reply tests/scripts/all.sh]: Sorry in my first comment I erroneously associated your comment to the wrong line.
Then, looking at the 2nd comment, I was like: "Ok, even if it builds, then wouldn't it be better to still disable all the `MBEDTLS` symbols for modes/key-pairs as well. They are not used so what's the purpose for keeping them there?". But then I tried and found that there were 3 components in `test_suite_psa_crypto` which [were failing](https://github.com/Mbed-TLS/mbedtls/pull/8119/commits/1ce6f0fabbfa15bce6c5a42f5cc799ee4ca7a422) due to wrong dependencies in the test. So this change was really beneficial in the end :)

Comment (mpg) [reply tests/scripts/all.sh]: I think as a general rule of thumb, it's always better to keep as many things enabled as possible, in order to get better test coverage. The obvious exception being tests that are focused on how much you can disable while keeping (some) things working :)

**Review by mpg — DISMISSED** [review]
LGTM, thanks!

File: tests/suites/test_suite_psa_crypto.data
Comment (mpg) [inline tests/suites/test_suite_psa_crypto.data:3204]: Note: I've created https://github.com/Mbed-TLS/mbedtls/issues/8415 as a small follow-up.

**Comment by valeriosetti** [comment]
Just rebased this PR in order to solve conflicts with `development`

**Review by mpg — APPROVED** [review]
Re-approving after rebase.

**Comment by mpg** [comment]
@gabor-mezei-arm Can you re-review this? (Or if you're maxed out with other things at the moment, let us know so that someone else can pick this up.)

**Review by gabor-mezei-arm — APPROVED** [review]
LGTM

**Comment by valeriosetti** [comment]
Sounds like InternalCI experienced some issue but the Open one did not. Can we still add this to the merge queue?

**Comment by mpg** [comment]
> Sounds like InternalCI experienced some issue but the Open one did not. Can we still add this to the merge queue?

Yeah, Internal has been pretty flaky lately. As a result, we've switched to Open being the mandatory one for the merge queue, so it should still work.

**Pull request merged**
